realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProgressListenerTests.java

```diff
@@ -83,19 +83,23 @@ private void assertTransferComplete(Progress progress, boolean nonZeroChange) {
     }
 
     // Create remote data for a given user.
-    private URI createRemoteData(SyncUser user) {
-        final SyncConfiguration config = configFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM)
-                .name("remote")
-                .build();
+    private URI createRemoteData(final SyncConfiguration config) {
         final Realm realm = Realm.getInstance(config);
         final CountDownLatch changesUploaded = new CountDownLatch(1);
         final SyncSession session = SyncManager.getSession(config);
         final long beforeAdd = realm.where(AllTypes.class).count();
         writeSampleData(realm);
 
+        final long threadId = Thread.currentThread().getId();
+
         session.addUploadProgressListener(ProgressMode.INDEFINITELY, new ProgressListener() {
             @Override
             public void onChange(Progress progress) {
+                // FIXME: This check is to make sure before this method returns, all the uploads has been done.
+                // See https://github.com/realm/realm-object-store/issues/581#issuecomment-339353832
+                if (threadId == Thread.currentThread().getId()) {
```

Comment:
Shouldn't this check be moved into the library implementation? If we have this bug, end users will also run into it?

Reply:
I think the API design is actually buggy, moving the checking to implementation will create other issues -- like user wants to get the progress made by other threads.

We really need to address https://github.com/realm/realm-object-store/issues/581#issuecomment-339353832

Reply:
Good point. This isn't about thread ID's but Realm versions. My bad